moosetools/moosetest/base/Differ.py

```diff
@@ -0,0 +1,54 @@
+#* This file is part of MOOSETOOLS repository
+#* https://www.github.com/idaholab/moosetools
+#*
+#* All rights reserved, see COPYRIGHT for full restrictions
+#* https://github.com/idaholab/moosetools/blob/main/COPYRIGHT
+#*
+#* Licensed under LGPL 2.1, please see LICENSE for details
+#* https://www.gnu.org/licenses/lgpl-2.1.html
+
+import io
+import copy
+import platform
+import logging
+from moosetools.base import MooseObject
+
+
+def make_differ(cls, controllers=None, **kwargs):
+    """
+    Create a `Differ` object given the *cls* with the `validObjectParams` of the *controllers*.
+
+    This function operates in the same fashion as `moosetools.base.make_runner`.
+    """
+    params = cls.validParams()
+    for ctrl in controllers or []:
+        params.add(ctrl.getParam('prefix'), default=ctrl.validObjectParams())
+    return cls(params, **kwargs)
+
+
+class Differ(MooseObject):
```

Comment:
I think there would be a fair amount of value to have a subclass of Differ that compares gold and output files, since there is a lot of commonality between all file differs (such as getting the gold file for a given input, checking for file existence etc).  So I would recommend making a "FileDiffer" class that is also part of moosetest.

Reply:
Agreed

Reply:
I was actually thinking of just creating a single FileDiffer that used the extension to call the correct diffing algorithm. But the problem with that approach is that tolerance settings, per file extension might be problematic. However, we could have "csv_abs_tol" and "json_abs_tol" parameters.

Reply:
Hm, in RAVEN we do have different differs that take the same file type.  For example we have an ordered CSV differ that checks that the numbers are the same and in the same order, and a different unordered CSV differ that just checks that all the entries are in the CSV file, but they can be in a different order.  We occasionally think about adding a L2 norm differ, which again, might very well take CSV files.

So I can see the use in automatically choosing the differ to use, but the opposite is quite useful sometimes.  In RAVEN, the Testers can call Differs directly, so maybe a Moose specific Runner could call the correct Differs based on filenames, but the general MooseTest would still have the flexibility to specify Differs.

Also, sometimes it is handy to be able to call the same differ with different parameters in the same test (for example, if we dump the input to an algorithm and also the output from it, we might be able to specify a much more accurate tolerance on the input.)